### PR TITLE
aerospace: add `workspace-to-monitor-force-assignment` option and fix `on-window-detected` type

### LINKED ISSUE #1142: `services.aerospace.settings.on-window-detected` has the wrong type
https://github.com/LnL7/nix-darwin/blob/189d2d422c773fa065cc9c72e6806f007ebb9be0/modules/services/aerospace/default.nix#L73

https://nikitabobko.github.io/AeroSpace/guide#on-window-detected-callback

it is not a list of strings; it is a set with atributes: { x = 1; y = 2; }

### PR DESCRIPTION
**Pull request by thuvasooriya**
trying to fix #1142

this might be redundant or even extra since i found out later that we already have #1168
fixes the error for most of my use cases not sure if there are any other issues.

let me know if there is some changes that i should make for this to be merged.

### CODE REVIEW

**Comment by thuvasooriya** [comment]
This one solves the issue #1142 and implements [workspace-to-monitor-force-assignment]
@Enzime

**Review by Enzime — CHANGES_REQUESTED** [review]
Can you remove all the formatting changes?

Can you update the tests as well?

File: tests/services-aerospace.nix
Comment (Enzime) [inline tests/services-aerospace.nix:32]: Can you update the settings above and use the new option you added and set `on-window-detected` maybe with the example value?

File: tests/services-aerospace.nix
Comment (Enzime) [inline tests/services-aerospace.nix:34]: Can you replace this check with something that actually checks the contents of `$conf` to see that it is setting the settings correctly like:

https://github.com/LnL7/nix-darwin/blob/c6b65d946097baf3915dd51373251de98199280d/tests/services-skhd.nix#L20

**Review by Enzime — CHANGES_REQUESTED** [review]

File: tests/services-aerospace.nix
Comment (thuvasooriya) [reply tests/services-aerospace.nix:34]: i was doubtful of the initial implementation of the testing as well, unfortunately I didn't know much about it, I'll try this weekend to replicate the skhd one.

**Comment by z0al** [comment]
Works locally for me. Thanks @thuvasooriya for taking care of this

File: modules/services/aerospace/default.nix
Comment (z0al) [inline modules/services/aerospace/default.nix:75]: Is there any reason for `types.anything` instead of a proper sub module?

Comment (talhaHavadar) [reply modules/services/aerospace/default.nix:75]: I did try to come up with something more restricted (not much really) in #1168 but don't have much knowledge about aerospace and possible values of this toml table so we kind of backed up to accept everything. Just leaving it here in case we want to play with it a bit.
```
  callbackConf =
    with lib.types;
    oneOf [
      bool
      int
      float
      str
      path
      attrs
    ]
    // {
      description = "AeroSpace config callback (bool, int, float, str, path or attrs)";
    };
...
            on-window-detected = lib.mkOption {
              type = listOf (attrsOf (either callbackConf (listOf callbackConf)));
...
```

Comment (z0al) [reply modules/services/aerospace/default.nix:75]: > don't have much knowledge about aerospace and possible values of this toml table

Looking at the code & docs I believe: 
1.`run` command is an array of strings and it's required.
2. `if.app-id`, `if.workspace`, `if.window-title-regex-substring` and `if.app-name-regex-substring` are all optional and of type string
3. `if.during-aerospace-startup` is a boolean and optional 
4. `check-further-callbacks` is optional and of type boolean

I will leave it to @thuvasooriya if they want to add proper typing here. Otherwise, I can follow up with a PR once this is merged.

Comment (thuvasooriya) [reply modules/services/aerospace/default.nix:75]: Thankyou for the valuable PR @talhaHavadar, and @z0al for taking the time to go through the docs. This was my first Nix contribution ever so I didn't want to mess around too much.
My initial solution before I saw #1168 was
```
              type = listOf (attrsOf (oneOf [
                bool
                str
                attrs
              ]));
```
which was not rigid in terms of types but then I changed it according to the suggestion made by @Enzime in #1168.
I'll see if I can come up with anything better. Thanks again.

Comment (thuvasooriya) [reply modules/services/aerospace/default.nix:75]: https://github.com/thuvasooriya/nix-darwin/commit/49c1990eab619be2fc46528cc2d797316585edb3
let me know if this can be merged to this PR @Enzime @z0al
Tried to whip something up from the above doc description, not sure if it handles all the edge cases. but works with the test cases and my config

Comment (thuvasooriya) [reply modules/services/aerospace/default.nix:75]: Here is some information, so that some smart people can figure out what's wrong quickly. Hope it helps. I've tried some methods to make the fields with default values disappear only to fail.

current PR generated config - **works properly**
```toml
[[on-window-detected]]
run = "layout floating"

[on-window-detected.if]
app-name-regex-substring = "telegram|finder"

[[on-window-detected]]
run = "move-node-to-workspace 0"

[on-window-detected.if]
app-name-regex-substring = "ripes"
```

my shoddy implementation of proper types in https://github.com/thuvasooriya/nix-darwin/commit/49c1990eab619be2fc46528cc2d797316585edb3 - **doesn't work**

```toml
[[on-window-detected]]
check-further-callbacks = true
run = "move-node-to-workspace 0"

[on-window-detected.if]
app-id = ""
app-name-regex-substring = "ripes"
during-aerospace-startup = false
window-title-regex-substring = ""
workspace = ""
```

Comment (Enzime) [reply modules/services/aerospace/default.nix:75]: If you change the type of `app-id` to be `types.nullOr types.str` and set the default value to `null`, it should omit it from the output

Comment (thuvasooriya) [reply modules/services/aerospace/default.nix:75]: > If you change the type of `app-id` to be `types.nullOr types.str` and set the default value to `null`, it should omit it from the output

Yeah that's one of the very first things i tried

```nix
                        app-id = lib.mkOption {
                          type = nullOr str;
                          default = null;
                          description = "The application ID to match (optional).";
                        };
```

```sh
error: builder for '/nix/store/6i98bgjlsmqsaqn5hz3id9p0ipchf3ln-aerospace.toml.drv' 
failed with exit code 1;
       last 1 log lines:
       > Error: Cannot convert data to TOML (null values are not supported)
```

Comment (z0al) [reply modules/services/aerospace/default.nix:75]: You can try to [filter the config](https://github.com/NixOS/nixpkgs/blob/0850b8d575c01f01d52084419fa5932046b29d7a/nixos/modules/services/security/kanidm.nix#L45C18-L45C68) of null values before passing the settings to`toml.generate`

Comment (thuvasooriya) [reply modules/services/aerospace/default.nix:75]: > You can try to [filter the config](https://github.com/NixOS/nixpkgs/blob/0850b8d575c01f01d52084419fa5932046b29d7a/nixos/modules/services/security/kanidm.nix#L45C18-L45C68) of null values before passing the settings to`toml.generate`

This is what i was looking for.. Thx!

Comment (thuvasooriya) [reply modules/services/aerospace/default.nix:75]: unfortunately replicating the filter is somewhat troublesome for me.
my guess is that the list of submodules is not playing nicely with filterAttrs.
I tried the following and still get the above cannot convert data to TOML, meaning the null is not filtered.

I tried the default functions, filterAttrs, filterAttrsRecursive
Then I tried to implement a custom function to iterate over each item in the submodule list. Which doesn't work either. Maybe there is some obvious mistake that I'm overlooking. 
Any help or direction would be appreciated. @Enzime @z0al
```nix
  filterOptionalFields = value:
    if lib.isAttrs value then
      let
        filteredAttrs = lib.mapAttrs (_: v: filterOptionalFields v) value;
        nonNullAttrs = lib.filterAttrs (_: v: v != null) value;
      in
        filteredAttrs // nonNullAttrs
    else if lib.isList value then
      lib.filter (x: x != null) (map filterOptionalFields value)
    else
      value;
  # filterOptionalFields = lib.filterAttrsRecursive (_: v: v != null);
  configFile = format.generate "aerospace.toml" (filterOptionalFields cfg.settings);
```

Comment (z0al) [reply modules/services/aerospace/default.nix:75]: Ah, I see. My bad. I didn't know that `filterAttrsRecursive` doesn't handle the case of lists.

Here is a custom implementation that works. I took the original `filterAttrsRecursive` code and added a condition for lists + changed the predicate signature to accept only values.

```nix
  # Code needs proper formatting
  filterAttrsRecursive =
    pred:
    set:
    lib.listToAttrs (
      lib.concatMap
        (name:
        let v = set.${name}; in
        if pred v then [
          (lib.nameValuePair name (
            if lib.isAttrs v then filterAttrsRecursive pred v
            else if lib.isList v then
              (
                map
                  (i:
                  if lib.isAttrs i then
                    filterAttrsRecursive pred i
                  else i)
                  (lib.filter pred v)
              )
            else v
          ))
        ] else [ ]
        )
        (lib.attrNames set)
    );
  filterNulls = filterAttrsRecursive (v: v != null);
  configFile = format.generate "aerospace.toml" (filterNulls cfg.settings);
```

Let's hope that Enzime has a trick up their sleeve so we don't have to keep this custom function :) 

P.S. Don't forget to change all optional attributes to default to null even things like `check-further-callbacks` as Aerospace behaves differently when it's set to `false` vs left out.

Comment (thuvasooriya) [reply modules/services/aerospace/default.nix]: cool! it works! thanks.
merged the changes, and yes every optional is set to null by default.

File: tests/services-aerospace.nix
Comment (z0al) [inline tests/services-aerospace.nix:72]: It would be great if you also check that unspecified options don't appear in the config e.g. the `window-title-regex-substring`.

Comment (thuvasooriya) [reply tests/services-aerospace.nix:72]: I don't think default null values have a way of somehow entering TOML files, since they have the concept of null = omitted. Hence I'm not sure how to feasibly test that in this specific implementation. maybe we check that `check-further-callbacks = false` is not present before the line `run = "layout floating"`

Comment (Enzime) [reply tests/services-aerospace.nix:72]: I think it's fine to do a negative assertion for fields that don't show up in the config file, you can also be explicit about the value by setting them to `null` intentionally rather than relying on the default value of the option

You can see how I did them here: https://github.com/LnL7/nix-darwin/commit/fd6660cb9182fde5e593246e311dc3c2dd4b9d13

Comment (thuvasooriya) [reply tests/services-aerospace.nix:72]: thanks for the hint!
I'll try to tackle that when I get some time.

**Comment by thuvasooriya** [comment]
@Enzime just made a small change to detect `window-title-regex-substring` not being present, and added example configs for all the other callback examples. if we have to do this properly then we should check how the whole file is output in order to check the presence or absence of certain troublemakers.

This works flawless for now. Let me know if you want anything else for this to be merged.

**Comment by yzx9** [comment]
Hello, just checking in—any updates? I really need this!

**Comment by thuvasooriya** [comment]
@Enzime pinging again just in case you missed this. I hope this has no blockers for a merge.

**Comment by Enzime** [comment]
Can you squash all the commits and rebase this on top of master

**Comment by thuvasooriya** [comment]
@Enzime can you confirm this is what you requested? never done this sort of thing before

**Review by Enzime — APPROVED** [review]
LGTM

Thanks for the contribution

**Pull request merged**
